### PR TITLE
Port diff_match_patch into Typescript

### PR DESCRIPTION
**Pull request by samelhusseini**
Ported the Javascript version of diff_match_patch into Typescript.
- [x] I ported the JavaScript tests. All tests passing.
- [x] I checked in the generated Typings declaration (dts)
- [x] I also checked in the compiled UMD binary for those wanting to directly import that into their projects.

To run tests:
```
cd typescript/tests/
tsc
node ./built/tests/diff_match_patch_test.js
node ./built/tests/speedtest.js
```

### CODE REVIEW

**Comment by rgbkrk** [comment]
How well do the types line up with `@types/diff-match-patch` to ease migration?

**Comment by samelhusseini** [comment]
They are mostly a match with the exception of Diff. I decided to match other language conversions here and make Diff a class. It’s neater that way and allows us to call the fields operation and text rather than use them by their indices.
This would make it harder to migrate tho for those that had been using @types and the JS version.

What are your thoughts Kyle?

**Comment by rgbkrk** [comment]
I'm not going to be migrating from the `@types` so it's not personally a blocker. From my purview, it seems ok to force a migration since the typescript here is still bundled in a big repo (no one can install this directly). However, it's reasonable to reach out to @karak, @pspeter3, and @vsiao to see what they think as prior authors of the old definitions.

**Comment by pspeter3** [comment]
I'm more worried about the breaking change to the JavaScript than the TypeScript. Having real guaranteed correct types (because the compiler generates the .d.ts) is worth the migration pain.

**Comment by karak** [comment]
Type definition in @types is just a mirror. I agree with the library itself has one and it's officially preferred.
I think the issue is the change in JavaScript after all.

**Comment by samelhusseini** [comment]
One thing we can do to help ease migration is support both types. 
[number, string] and the Diff class. 

[Here's](https://www.typescriptlang.org/play/index.html#code/KYDwDg9gTgLgBAYwgOwM7wCIEkBiOD6WAcgMoCiASgCpwC8cAjANwBQLokscMAnmMHCoBXMABsB9ANrIhAWwBGwKABo46KAEtkAcwC6rduGjxe-OFgwaAZlbqCR4uAB84lmwY7HEogIapUrta2AN4scHAADABccDIKSqzhDDHqWtoG4UhoMFBCCDDQABQQYDFxiircoDApOWkAlHCh4eEwABYaqJIRunYliS3tnZIMvfQw1QMAvmzh2sDwJUo+MBoohfVlchVNYS1QC0JQyNwdXT1wfrHbCXsze6gLcEtQK2vIxaXX8VCNzYNnbpjZ5gaazODzEzVDa1TQ6XYtOAHGBHE5DLqjS4BVI6MEPJ4TEAwQqEmpqOo6P57VqAzHjSZ3cEAegAVCzqSy4GRZEJfBNTgIIEIYGBhc9bD5uAB3CAAWmA4lkwGQ8B8UFePAAdBy4AABZGopo47RTQI2EHLVYoLGXclw7Talospl7AokCnaDYI-aHY6nYYXADUcAARMoQ3Bg+iRvpGfcsugqon6IUACZBGIWIKNWgAPm9Br96ZsmrdHo2rCmBgT8GLtnoyGAUrNVkK2DwhFIlCoqhDDBD9VYNbgdYATJm3PW4N1e-3Yyxh3WqMBk0niXXBwuUImlyuYKO7BN0GmgqPN1vsiC7HXNS83igh9vFmAD-Q60Dq0+qkTr0FSwzh1JV8RyCGNWCAA) a quick prototype of my suggestion.

tl;dr. 
- Methods that accept a diff, use IDiff instead (which is a union type of [number,string] and the Diff class)
- For folks that have been using new Diff(number,string) but have been referencing [0] and [1], we can use getters and setters to wrap operation and text in the Diff class. 

--- Edit
As I think about this more, I don't think the above would work as you could end up having defined a Diff as an array. [1, 'text'] and error trying to access its operation field.

I think it makes more sense to just provide helpers for anyone coming from the JS world. 
If you've defined your Diffs as arrays, perhaps we could defined a factory method on Diff ``.fromArray()`` and if you're accessing a diff object's operation via the index or via operation, the above should getters / setters should handle that. Like [so](https://www.typescriptlang.org/play/index.html#code/KYDwDg9gTgLgBAYwgOwM7wCIEkBiOD6WAcgMoCiASgCpwC8cAjANwBQLoksiANgIaqo4GAJYAzUXADeLOHAAMALjjIArgFsARsCitZDJeijDkAc1YzEKQyoQxoACghglqzdoA0cGKBgGYR0wBKKQtZGAALYVQAbTkAXTo4J11ZL0iYhgT6bxAYFIBfNll0XhhhBDhRKAg1AEEoKF4AT3sAEzFResampWjXLShPQ2MTOODpVLgoYBgVKGRlYAB3IQ62jq7m2LjPdvFNpujMwIKiuBMZpLBtUuEUe0CXdQGQyenZ+bSo7bh+ZWftKcLKhLk4bmV7k4nm4oONQl8YvFEskLIULBd4DkYA8-AETK9Uu85gsIt9Mr9BMNTEDipcsfYsbiRnDJqSMlkvD4aXAAPQAKj5oT5cDIahUfG8aWASRUMDAsqSEl4XiWEAAtMBuMA1MBkPBeA1mgA6IVwAACRM+kipJnyq3EV3BdwWf2VNpNqT5PIsdhI-hGDwJsktJPSPwA1HAAETuKNwSNso5xU5opBoTHAdCJdbiJQicTBWgAPgJIbge1ERt9-tMD1Y+XMaazFcSyGW9tE9mweEIpEoVE8UYYUZOLCb8ArACZEvnK1Uagd7NFuwRiORqIPh2NG1YJx0qJn4NlDznRKPx+X94fp8f0KfJ6OWBWjWDGhDkKwK9tzM+sZ+OkmP4dJOL7XG+zr-uIk7fmwU5VlyT7AYBLBAA)

**Review by rgbkrk — APPROVED** [review]

**Comment by karak** [comment]
Hey. Wasn't this issue closed, or unmerged yet?

**Comment by dmsnell** [comment]
Still unmerged. Things move slow here. You can tell because it says "Open" and not "Closed" or "Merged" 😉

It'd be nice if this were easier to review, but if it's up-to-date and all the tests pass it could be worth merging for no other reason than to keep it moving. Things I know I would like to see with this:

 - Incorporation of some kind of fix for the split surrogate pair issue in `toDelta`. See #69 #80
 - Some discussion about whether this work could also eliminate the role that ClosureCompiler plays, or if we have to stick with that for now.
 - Updated docs on the role `tsc` plays and how to build the project. Currently those are on the Wiki which is admittedly hard to track in source control.
 - Deletion of the old `_uncompressed` JS.
